File: clover-android-sdk-examples/src/main/java/com/clover/android/sdk/examples/PrintJobsTestActivity.java
/**
 * Copyright (C) 2015 Clover Network, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.clover.android.sdk.examples;

import android.accounts.Account;
import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.text.TextUtils;
import android.util.Pair;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import com.clover.sdk.util.CloverAccount;
import com.clover.sdk.v1.printer.Category;
import com.clover.sdk.v1.printer.Printer;
import com.clover.sdk.v1.printer.PrinterConnector;
import com.clover.sdk.v1.printer.job.PrintJob;
import com.clover.sdk.v1.printer.job.PrintJobsConnector;
import com.clover.sdk.v1.printer.job.PrintJobsContract;
import com.clover.sdk.v1.printer.job.TestReceiptPrintJob;

import java.util.Arrays;
import java.util.List;

public class PrintJobsTestActivity extends Activity {
  private static final String TAG = PrintJobsTestActivity.class.getSimpleName();

  private static final Handler uiHandler = new Handler(Looper.getMainLooper());

  private Account account;
  private PrinterConnector printerConnector;

  private EditText editPrinterId;
  private Button buttonPrint;
  private TextView textId;

  private Button buttonGetPrintJobIds;
  private TextView textPrintJobIds;

  private HandlerThread handlerThread = new HandlerThread(this.getClass().getName());
  private Handler handler;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_printjobs_test);

    handlerThread.start();
    handler = new Handler(handlerThread.getLooper());

    account = CloverAccount.getAccount(this);
    printerConnector = new PrinterConnector(this, account, null);
    fillPrinterId();

    textId = (TextView) findViewById(R.id.text_id);

    editPrinterId = (EditText) findViewById(R.id.edit_printer_id);
    buttonPrint = (Button) findViewById(R.id.button_print);
    buttonPrint.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        new AsyncTask<Void, Void, String>() {
          @Override
          protected String doInBackground(Void... voids) {
            try {
              Printer p = getPrinter();
              PrintJob printJob = new TestReceiptPrintJob.Builder().build();
              return new PrintJobsConnector(PrintJobsTestActivity.this).print(p, printJob);
            } catch (Exception e) {
              e.printStackTrace();
            }
            return null;
          }

          @Override
          protected void onPostExecute(String id) {
            if (id == null) {
              return;
            }
            textId.setText(id);
            monitorState();
          }
        }.execute();
      }
    });

    textPrintJobIds = (TextView) findViewById(R.id.text_printjob_ids);
  }

  private void monitorState() {
    handler.postDelayed(new Runnable() {
      @Override
      public void run() {
        PrintJobsConnector connector = new PrintJobsConnector(PrintJobsTestActivity.this);

        final List<String> inQueueIds = connector.getPrintJobIds(PrintJobsContract.STATE_IN_QUEUE);
        final List<String> printingIds = connector.getPrintJobIds(PrintJobsContract.STATE_PRINTING);
        final List<String> doneIds = connector.getPrintJobIds(PrintJobsContract.STATE_DONE);
        final List<String> errorIds = connector.getPrintJobIds(PrintJobsContract.STATE_ERROR);

        final StringBuilder printJobIds = new StringBuilder();

        List<Pair<String,List<String>>> pairs = Arrays.asList(
            Pair.create("In queue: ", inQueueIds),
            Pair.create("Printing: ", printingIds),
            Pair.create("Done: ", doneIds),
            Pair.create("Error: ", errorIds)
        );
        for (Pair<String,List<String>> pair: pairs) {
          printJobIds.append(pair.first);
          for (String id: pair.second) {
            printJobIds.append(id).append(", ");
          }
          if (!pair.second.isEmpty()) {
            printJobIds.delete(printJobIds.length() - 2, printJobIds.length());
          }
          printJobIds.append("\n\n");
        }
        uiHandler.post(new Runnable() {
          @Override
          public void run() {
            textPrintJobIds.setText(printJobIds.toString());
          }
        });

        monitorState();
      }
    }, 100);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();

    uiHandler.removeCallbacksAndMessages(null);
    handler.removeCallbacksAndMessages(null);

    if (printerConnector != null) {
      printerConnector.disconnect();
      printerConnector = null;
    }
  }

  private void fillPrinterId() {
    new AsyncTask<Void,Void,List<Printer>>() {
      @Override
      protected List<Printer> doInBackground(Void... params) {
        try {
          return printerConnector.getPrinters(Category.RECEIPT);
        } catch (Exception e) {
          e.printStackTrace();
        }
        return null;
      }

      @Override
      protected void onPostExecute(List<Printer> printers) {
        if (printers != null && !printers.isEmpty()) {
          editPrinterId.setText(printers.get(0).getUuid());
        }
      }
    }.execute();
  }

  private Printer getPrinter() {
    String id = editPrinterId.getText().toString();
    if (TextUtils.isEmpty(id)) {
      try {
        List<Printer> printers = printerConnector.getPrinters(Category.RECEIPT);
        if (printers != null && !printers.isEmpty()) {
          return printers.get(0);
        }
      } catch (Exception e) {
        e.printStackTrace();
      }
      return null;
    }

    try {
      return printerConnector.getPrinter(id);
    } catch (Exception e) {
      e.printStackTrace();
    }
    return null;

  }
}

File: clover-android-sdk-examples/src/main/java/com/clover/android/sdk/examples/PrintJobsTestActivity.kt
/**
 * Copyright (C) 2016 Clover Network, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.clover.android.sdk.examples

import android.accounts.Account
import android.database.ContentObserver
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.clover.android.sdk.examples.databinding.ActivityPrintjobsTestBinding
import com.clover.sdk.util.CloverAccount
import com.clover.sdk.v1.printer.Category
import com.clover.sdk.v1.printer.Printer
import com.clover.sdk.v1.printer.PrinterConnector
import com.clover.sdk.v1.printer.job.PrintJobsConnector
import com.clover.sdk.v1.printer.job.PrintJobsContract.AUTHORITY_URI
import com.clover.sdk.v1.printer.job.PrintJobsContract.STATE_DONE
import com.clover.sdk.v1.printer.job.PrintJobsContract.STATE_ERROR
import com.clover.sdk.v1.printer.job.PrintJobsContract.STATE_IN_QUEUE
import com.clover.sdk.v1.printer.job.PrintJobsContract.STATE_PRINTING
import com.clover.sdk.v1.printer.job.TestReceiptPrintJob
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * This class is an example usage of the [PrintJobsConnector] class. It shows two buttons:
 * - Print: This looks up a configured receipt printers and calls [PrintJobsConnector.print] (with a printer argument).
 * - Print to any: Calls [PrintJobsConnector.print] (without a printer argument).
 *
 * It then observes the print jobs and displays their state. For example, "print" is
 * clicked expect the print job to move from "in queue" to "printing" and finally "done".
 */
open class PrintJobsTestActivity : AppCompatActivity() {

  private lateinit var account: Account
  private lateinit var printerConnector: PrinterConnector
  private lateinit var printJobsConnector: PrintJobsConnector
  private lateinit var binding: ActivityPrintjobsTestBinding

  private val printJobsObserver = object : ContentObserver(Handler(Looper.getMainLooper())) {
    override fun onChange(selfChange: Boolean) {
      lifecycleScope.launch { updateState() }
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    binding = ActivityPrintjobsTestBinding.inflate(layoutInflater)
    setContentView(binding.root)

    account = CloverAccount.getAccount(this)
    printerConnector = PrinterConnector(this, account, null)
    printJobsConnector = PrintJobsConnector(this)

    binding.buttonPrint.setOnClickListener {
      lifecycleScope.launch {
        withContext(Dispatchers.IO) {
          receiptPrinter?.let { p ->
            printJobsConnector.print(p, TestReceiptPrintJob.Builder().build())
          }
        }
      }
    }
    binding.buttonPrintToAny.setOnClickListener {
      lifecycleScope.launch {
        withContext(Dispatchers.IO) {
          printJobsConnector.print(TestReceiptPrintJob.Builder().build())
        }
      }
    }
  }

  override fun onStart() {
    super.onStart()
    // Observe print job changes
    contentResolver.registerContentObserver(AUTHORITY_URI, true, printJobsObserver)

    // Populate initial values
    lifecycleScope.launch { updateState() }

    // Do we have a receipt printer configured?
    lifecycleScope.launch {
      withContext(Dispatchers.IO) { receiptPrinter }?.let {
        binding.layoutButtons.visibility = View.VISIBLE
        binding.textError.visibility = View.GONE
      } ?: run {
        binding.layoutButtons.visibility = View.GONE
        binding.textError.visibility = View.VISIBLE
      }
    }

  }

  override fun onStop() {
    contentResolver.unregisterContentObserver(printJobsObserver)
    super.onStop()
  }

  override fun onDestroy() {
    super.onDestroy()
    printerConnector.disconnect()
  }

  private suspend fun updateState() {
    val states = withContext(Dispatchers.IO) {
      mapOf(
          ::STATE_IN_QUEUE.name to printJobsConnector.getPrintJobIds(STATE_IN_QUEUE),
          ::STATE_PRINTING.name to printJobsConnector.getPrintJobIds(STATE_PRINTING),
          ::STATE_DONE.name to printJobsConnector.getPrintJobIds(STATE_DONE),
          ::STATE_ERROR.name to printJobsConnector.getPrintJobIds(STATE_ERROR),
      )
    }

    binding.printjobIds.text = StringBuilder().apply {
      for (entry in states) {
        append("${entry.key.padEnd(16)}: ")
        for (id in entry.value) {
          append(id)
          if (id != entry.value.last()) {
            append(", ")
          }
        }
        append("\n\n")
      }
    }
  }

  /**
   * Return a [Printer] for the category [Category.RECEIPT], or `null` if no such
   * printer is configured. If there are multiple category [Category.RECEIPT]
   * printers configured one is selected at random.
   */
  private val receiptPrinter: Printer?
    get() = printerConnector.getPrinters(Category.RECEIPT)?.let { printers ->
      if (printers.isEmpty()) {
        return null
      }
      return printers[0]
    } ?: run { return null }
}